### ISSUE TITLE
IDL-level support for permission checks?

### ISSUE THREAD

**Issue by marcoscaceres** [whatwg/webidl#1066]
Not sure if this will work, but thought I'd pitch it... 

Methods defined by various specs will check in a method call is "allowed to use" a particular permission. What might be nice is to have IDL level support for this. I was thinking something like an extended attribute, like: 

```
partial interface Navigator {
  [Permission="webshare"] Promise<undefined> share(optional ShareData data = {});
  [Permission="webshare"] boolean canShare(optional ShareData data = {});
};
```

In cases where the return type is a boolean, it would return `false` if not allowed - otherwise, do whatever the steps of the method require.

In other cases, it would throw or reject with a "NotAllowedError" DOMException. 

cc @dontcallmedom @tidoust

**Comment by annevk** [comment]
This could maybe work similar to https://html.spec.whatwg.org/#cereactions, though the IDL steps (method steps, etc.) don't have access to the types involved. So you'd need something like `[Permission=(Promise<undefined>, "webshare")]` I think.

(FWIW, I looked at Storage and Notifications API and neither could use this I think.)

**Comment by domenic** [comment]
This is similar to something we drafted in https://wicg.github.io/nav-speculation/prerendering.html#delay-while-prerendering . If we can find enough specs that would use this then I'd be supportive. I suspect it might involve technically-backward-incompatible changes to exception ordering.

I would prefer starting only with the Promise variant, although gathering data on what specs would use this might change my mind there.

I don't understand why the type would need to be included in the arguments. I think you could just write in the spec for [Permission] "If the return type of the operation is X, do Y."

**Comment by annevk** [comment]
Yeah, I guess that's correct. It would have sufficient context. (I was thinking from the perspective of "method steps" and such, which don't get told the type.)

**Comment by domenic** [comment]
So to be clear I think the next step would be to (somehow) find all the callers of "allowed to use", and divide them up into categories like:

- Could use this with no modifications
- Could use this but it would change exception ordering
- Could not use this
- Classify each as promise vs. exception vs. boolean

**Comment by saschanaz** [comment]
Callers found from webref (except WICG ones, for now):

Promise rejection

* [`HTMLMediaElement#setSinkId()`](https://w3c.github.io/mediacapture-output/#dom-htmlmediaelement-setsinkid)
* [`Navigator#getBattery()`](https://w3c.github.io/battery/#ref-for-index-term-allowed-to-use-1)
* [`WakeLock#request()`](https://w3c.github.io/screen-wake-lock/#the-request-method)
* [`Navigator#share()`](https://w3c.github.io/web-share/#share-method)
* [`PublicKeyCredential#isUserVerifyingPlatformAuthenticatorAvailable()`](https://w3c.github.io/webauthn/#ref-for-allowed-to-use) 
* [`Element#requestFullscreen()`](https://fullscreen.spec.whatwg.org/#ref-for-allowed-to-use) (TypeError): 
* [`Navigator#requestMediaKeySystemAccess()`](https://w3c.github.io/encrypted-media/#dom-navigator-requestmediakeysystemaccess) (SecurityError): 
* [`Navigator#getGamePads()`](https://w3c.github.io/gamepad/#dom-navigator-getgamepads) (SecurityError): 
* [`HTMLVideoElement#requestPictureInPicture()`](https://w3c.github.io/picture-in-picture/#ref-for-allowed-to-use) (SecurityError): 
* [`Navigator#requestMIDIAccess()`](https://webaudio.github.io/web-midi-api/#dfn-algorithm-to-request-midi-access) (SecurityError)
* [`Document#requestStorageAccess()`](https://privacycg.github.io/storage-access/#ref-for-allowed-to-use) (No error object?): 

Exception
* [Constructors of Generic Sensors](https://w3c.github.io/sensors/#ref-for-allowed-to-use%E2%91%A2) (SecurityError, e.g. in [new AmbientLightSensor()](https://w3c.github.io/ambient-light/#ref-for-check-sensor-policy-controlled-features))
* [`new PaymentRequest()`](https://w3c.github.io/payment-request/#dfn-paymentrequest-paymentrequest) (SecurityError)

boolean
* [`Navigator#canShare()`](https://w3c.github.io/web-share/#canshare-data-method)

**Comment by dontcallmedom** [comment]
since this didn't include `getUserMedia`, I ran my own check in webref (in ed/links):
```sh
for i in `grep -l '"allowed-to-use"' *` ; do jq '.spec.url' $i|grep -v wicg ; done
```

This adds in the Promise rejection category:
* [MediaDevices#getUserMedia()](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediadevices-getusermedia) (`NotAllowedError` - although it can throw an `InvalidStateError` if the tab is not fully active and focused)
* [MediaDevices#getDisplayMedia()](https://w3c.github.io/mediacapture-screen-share/#dom-mediadevices-getdisplaymedia) (`NotAllowedError` - although it can throw an `InvalidStateError` due to transient activation check)

And in the Exception category:
* [ML.createContext()](https://webmachinelearning.github.io/webnn/#dom-ml-createcontext) (`SecurityError`) (not shipped anywhere yet though)

There are 8 WICG specs that links to "allowed to use":
* https://wicg.github.io/conversion-measurement-api/
* https://wicg.github.io/floc/
* https://wicg.github.io/idle-detection/
* https://wicg.github.io/keyboard-map/
* https://wicg.github.io/page-lifecycle/
* https://wicg.github.io/serial/
* https://wicg.github.io/webhid/
* https://wicg.github.io/web-otp/

**Comment by marcoscaceres** [comment]
Ok, we had a look through the lists above and this seems potentially feasible. 

There are some cases we should discuss, in particular, a few APIs do "fully active" checks **before** "allowed to use" check (e.g., [Fullscreen API](https://fullscreen.spec.whatwg.org/#ref-for-allowed-to-use), and [WebShare](https://w3c.github.io/web-share/#share-method)). 

The other question is: what exception should be thrown? SecurityError seems pretty popular, but maybe that can be configurable? (tho that's it's a bit ugly). 

The last interesting one is for booleans operations... can it be safe to assume that "not allowed" would always return false... There is only the one case we have for Web Share. 

The other small thing we noticed is that some editors are confusing getting "permission to use X" (Permissions API) with Permissions Policy (i.e., "allowed to use")... this might help there too.
